File: src/main/java/com/github/pauldambra/moduluschecker/chain/checks/DoubleAlternateCheck.java
package com.github.pauldambra.moduluschecker.chain.checks;

import com.github.pauldambra.moduluschecker.ModulusCheckParams;
import com.github.pauldambra.moduluschecker.valacdosFile.WeightRow;

import java.util.function.Function;

public final class DoubleAlternateCheck {

    public Boolean check(ModulusCheckParams params, Function<ModulusCheckParams, WeightRow> rowSelector) {
        WeightRow selectedRow = rowSelector.apply(params);

        int total = ModulusTotal.calculateDoubleAlternate(params.account, selectedRow.getWeights());

        if (selectedRow.isException(1)) {
            total += 27;
        }

        final int remainder = total % 10;
        return remainder == 0;
    }
}

File: src/main/java/com/github/pauldambra/moduluschecker/chain/checks/DoubleAlternateCheck.kt
package com.github.pauldambra.moduluschecker.chain.checks

import com.github.pauldambra.moduluschecker.account.BankAccount
import com.github.pauldambra.moduluschecker.valacdosFile.WeightRow


class DoubleAlternateCheck {
    fun check(
      bankAccount: BankAccount,
      weightRow: WeightRow): Boolean {

        val exceptionModifier = if (weightRow.hasException(1)) 27 else 0

        var total = ModulusTotal.calculateDoubleAlternate(bankAccount, weightRow.weights) + exceptionModifier

        return total % 10 == 0
    }
}